### ISSUE TITLE
Report in #517 of TimeoutPolicy causing hangs when used with HttpClientFactory

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#518]
**Summary**: Report by @toddmeinershagen in #517 of `TimeoutPolicy` causing hangs when used with HttpClientFactory

**Expected behavior**: `TimeoutPolicy` with `HttpClientFactory` does not cause hangs when used with HttpClientFactory

**Reported behaviour**: `TimeoutPolicy` with `HttpClientFactory`  causes hangs when used with HttpClientFactory

**Steps / Code to reproduce the problem**: not supplied.

**Comment by reisenberger** [comment]
@toddmeinershagen Thanks for raising this.  I cannot reproduce the issue stated in #517:

>adding a Policy.TimeoutAsync() to HttpClientFactory (from the AddPolicy() extension method in Microsoft.Extensions.Http.Polly library) 
>does not allow the HttpClient to do its work and always ends up timing out.

If I use the code posted below (run simply as a .NET Core 2.1 Console Application), the execution does not hang.  

Conversely, if I switch `client.BaseAddress` (in the below code) to some invalid URL and set, eg, `timeoutPolicy = Policy.TimeoutAsync<HttpResponseMessage>(TimeSpan.FromMilliSeconds(1));`, I get a timeout as expected.

Can you identify anything different between the code you are using, and the below sample? Or: does the sample below hang for you? Or: Are you able to provide repro code for the issue you are seeing? 

```
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Polly;

namespace PollyWithHttpClientFactoryInAConsole
{
    class Program
    {
        static async Task Main(string[] args)
        {
            var builder = new HostBuilder()
                .ConfigureServices((hostContext, services) =>
                {
                    IAsyncPolicy<HttpResponseMessage> timeoutPolicy = Policy.TimeoutAsync<HttpResponseMessage>(TimeSpan.FromSeconds(30));

                    services.AddHttpClient("GoogleClient", client =>
                    {
                        client.BaseAddress = new Uri("https://www.google.com");
                    }).AddPolicyHandler(timeoutPolicy);
                        
                    services.AddSingleton<IHostedService, ConsoleService>();
                });

            await builder.RunConsoleAsync();
        }

    }

    public class ConsoleService : IHostedService
    {
        private IHttpClientFactory _httpClientFactory;
        public ConsoleService(IHttpClientFactory httpClientFactory)
        {
            _httpClientFactory = httpClientFactory;
        }

        public async Task StartAsync(CancellationToken cancellationToken)
        {
            await MakeRequestsToRemoteService();
        }

        public async Task MakeRequestsToRemoteService()
        {
            HttpClient httpClient = _httpClientFactory.CreateClient("GoogleClient");

            HttpResponseMessage response;
            try
            {
                response = await httpClient.GetAsync("/");
            }
            catch (Exception e)
            {
                Console.WriteLine("Failed with exception: " + e);
                return;
            }

            if (!response.IsSuccessStatusCode)
            {
                Console.WriteLine("Failed with status code: " + response.StatusCode);
            }
            else
            {
                Console.WriteLine("Succeeded with status code: " + response.StatusCode);
            }
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            return Task.CompletedTask;
        }
    }
}
```

**Comment by toddmeinershagen** [comment]
@reisenberger Thanks for responding so quickly.  I also appreciate your code.  There are a couple of differences I see with your code.  I was not running my code in a IHostedService.  I was merely new-ing up a ServiceCollection and loading the logging and polly policies.

I also am using multiple policies.  Before the Timeout policy, I have added a retry policy and a circuitbreaker policy.  The other difference is that I am using a timeout policy with a smaller period (10 seconds) and I am calling to a service (Http-based Azure Function hosted locally) that is starting up at the same time and times out on the first try.  It seems like once it fails the first try, the rest are continually failing.  (I am thinking that is what may be tripping the rest of the calls, but I will need more time to experiment.)

Right now - the code is not in a single file format, but I could zip it up if you wanted me to send it to you.

**Comment by toddmeinershagen** [comment]
@reisenberger Sorry it took my a little bit to get back to you.  The exception that bubbles up is the TimeoutRejectedException that is thrown from the Timeout policy.  The Retry policy is set up to handle those, but after the last retry, it will bubble out.

I was able to simplify the solution, however, I don't think I can get this same condition to occur with calls to Google.  I would have to have the only the first try to timeout.  Since I don't control the performance of the Google site, that variable would be hard to control.  (You may have some ideas about how to create that condition.)

After having worked with this for a bit, I think the main difference is that I am calling to a local v2 Azure Function.  It still doesn't make sense why that would cause it to consistently fail when using the Timeout policy, but for some reason, it appears that it never is able to make calls to the function even after the function is ready to receive Http triggered traffic.  

I even tried starting the Azure Function first and then detaching.  When I ran the client application, it still is not able to call out to the Azure function.  I don't believe the same occurs if I am calling an ASP.NET core API.  So, this may have something to do with the calls to the local Azure Function.  Not sure why the Timeout policy would be impacted by this.  

I have attached the simplified solution to see if you are able to reproduce the issue on your machine.
[timeouts-broken.zip](https://github.com/App-vNext/Polly/files/2494771/timeouts-broken.zip)

**Comment by reisenberger** [comment]
@toddmeinershagen The configured `TimeoutPolicy` allows `TimeSpan.FromSeconds(1)` for the timeout.  What happens if you lengthen that timeout, to - just for the sake of testing - 30 seconds?  When you have the `TimeoutPolicy` commented out, how long does the call take to succeed?

**Comment by toddmeinershagen** [comment]
@reisenberger You are correct that it works at 30 seconds.  I originally had it at 10 seconds and shortened it during the simplification so you could see the issue quicker.  Do you have any idea why if the timeout happens on the first time below a certain limit that the rest of the retries never make connection?

**Comment by reisenberger** [comment]
@toddmeinershagen What happens if you set the timeout policy as follows? This policy will timeout at 1 second the first time, then allow a 30-second timeout for subsequent tries.

        private static IAsyncPolicy<HttpResponseMessage> GetSharedTimeoutPolicy()
        {
            return Policy
                .TimeoutAsync<HttpResponseMessage>(
                    ctx => 
                        !ctx.ContainsKey("subsequent") 
                        ? TimeSpan.FromSeconds(1) 
                        : (TimeSpan)(ctx["subsequent"] = TimeSpan.FromSeconds(30)), 
                onTimeoutAsync: async (context, timeoutPeriod, task) =>
                {
                    await Console.Out.WriteLineAsync($"timeout ({timeoutPeriod})");
                });
        }

**Comment by toddmeinershagen** [comment]
@reisenberger Thanks for the suggestion.  I had to modify the policy slightly to make sure that when the key didn't exist, it would set the subsequent context item.  Otherwise, it failed every time because it only used the 1 second timeout.

After running it against Google, it appears that the first failure does not limit it from successfully running going forward on retries.  There must be something specific to Azure functions running locally.  Perhaps the http trigger mechanism gets backed up with the initial requests and blocks other requests from being passed to the function on retries.

Below is the sample code.

```csharp
class Program
    {
        static async Task Main(string[] args)
        {
            var builder = new HostBuilder()
                .ConfigureServices((hostContext, services) =>
                {
                    var timeoutPolicy = GetSharedTimeoutPolicy();
                    var retryPolicy = GetSharedRetryPolicyWithJitter();

                    services.AddHttpClient("GoogleClient", client =>
                    {
                        client.BaseAddress = new Uri("https://www.google.com");
                    })
                        .AddPolicyHandler(retryPolicy)
                        .AddPolicyHandler(timeoutPolicy);

                    services.AddSingleton<IHostedService, ConsoleService>();
                });

            await builder.RunConsoleAsync();
        }

        private static IAsyncPolicy<HttpResponseMessage> GetSharedTimeoutPolicy()
        {
            return Policy
                .TimeoutAsync<HttpResponseMessage>(ctx =>
                    {
                        var key = "subsequent";

                        if (ctx.ContainsKey(key))
                        {
                            return (TimeSpan) ctx[key];
                        }

                        ctx[key] = TimeSpan.FromSeconds(5);
                        return TimeSpan.FromMilliseconds(1);
                    },
                    onTimeoutAsync: async (context, timeoutPeriod, task) =>
                    {
                        await Console.Out.WriteLineAsync($"timeout ({timeoutPeriod})");
                    });
        }

        private static IAsyncPolicy<HttpResponseMessage> GetSharedRetryPolicyWithJitter()
        {
            return HttpPolicyExtensions
                .HandleTransientHttpError()
                .Or<TimeoutRejectedException>()
                .RetryAsync(
                    retryCount: 3,
                    onRetryAsync: async (e, i) => await Console.Out.WriteLineAsync($"retry attempt:  {i}"));
        }
    }

    public class ConsoleService : IHostedService
    {
        private readonly IHttpClientFactory _httpClientFactory;
        public ConsoleService(IHttpClientFactory httpClientFactory)
        {
            _httpClientFactory = httpClientFactory;
        }

        public async Task StartAsync(CancellationToken cancellationToken)
        {
            await MakeRequestsToRemoteService();
        }

        public async Task MakeRequestsToRemoteService()
        {
            HttpClient httpClient = _httpClientFactory.CreateClient("GoogleClient");

            HttpResponseMessage response;
            try
            {
                response = await httpClient.GetAsync("/");
            }
            catch (Exception e)
            {
                Console.WriteLine("Failed with exception: " + e);
                return;
            }

            if (!response.IsSuccessStatusCode)
            {
                Console.WriteLine("Failed with status code: " + response.StatusCode);
            }
            else
            {
                Console.WriteLine("Succeeded with status code: " + response.StatusCode);
            }
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            return Task.CompletedTask;
        }
    }
```

**Comment by reisenberger** [comment]
Thanks @toddmeinershagen for seeing the intent and making the obvious fix. (I was timing-slicing to get you quick suggestions)

I am going to close out this item as we have demonstrated that neither `TimeoutPolicy` nor `HttpClientFactory` is implicated. (Never in doubt in my mind given all `TimeoutPolicy` does is [execute an action with a fresh timing-out `CancellationToken`](https://github.com/App-vNext/Polly/blob/3e03c896b8060d237de7358c986886fa4240f226/src/Polly.Shared/Timeout/TimeoutEngineAsync.cs#L22-L34) while `HttpClientFactory` [simply executes the inner `base.SendAsync(...)` of the `DelegatingHandler` through the policy](https://github.com/aspnet/HttpClientFactory/blob/0677ac9f1e3efeb7557291ed42c726c60bde9c93/src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs#L130); but it helps to demonstrate.)  

To answer an earlier question:

>Since I don't control the performance of the Google site, [a timeout there] would be hard to control. (You may have some ideas about how to create that condition.)

A simple option could be to append a `DelegatingHandler` into the HttpClient pipeline which blocks for a configurable amount of time. (We're actually looking at formalising this as a [fault-injection policy in Polly](https://github.com/App-vNext/Polly/issues/499#issuecomment-421835612).) Or set up a simple WebAPI endpoint which blocks, as we do in [Polly-Samples](https://github.com/App-vNext/Polly-Samples/blob/2e7187b4eb3bca4f3351aef24f75f4156e8ec009/PollyTestApp/Controllers/NonThrottledFaultingController.cs),

**Comment by reisenberger** [comment]
@toddmeinershagen Haven't time to join you digging deeper at mo into the local Azure Function cold-start issues, unfortunately (time must go to other Polly issues), but interested what you find if you want to report back later.  Thoughts were that you might see if the same issue reproduces when the Azure function is run out of (say) a different Visual Studio instance; or indeed simply run up on Azure rather than locally.

**Comment by toddmeinershagen** [comment]
@reisenberger - I appreciate the dialog and ideas.  I did try your suggestion of running the Azure function in a separate instance of Visual Studio and got the same results.  However, when I published the function to the cloud and ran the client locally, it ran fine.

My guess is that there is some kind of optimization going on when running locally.  The interesting thing is that the call never makes it to the function.  So, whatever mechanics go on to handle the webhooks triggering do not work well in that scenario.
